### PR TITLE
Add BEBLID local feature descriptor

### PR DESCRIPTION
**Pull request by iago-suarez**
**Merge with extra**: https://github.com/opencv/opencv_extra/pull/827

BEBLID is a efficient binary descriptor learned with boosting. In several benchmarks it has proved to largely improve other binary descriptors like ORB or BRISK with the same efficiency. BEBLID describes using the difference of mean gray values in
different regions of the image around the KeyPoint, the descriptor is specifically optimized for
image matching and patch retrieval addressing the asymmetries of these problems.

<cut/>

image

This Pull Request ports the [original code](https://github.com/iago-suarez/BEBLID) to OpenCV. The article explaining all the details and comparing the descriptor in several tasks can be found [here](https://raw.githubusercontent.com/iago-suarez/BEBLID/master/BEBLID_Boosted_Efficient_Binary_Local_Image_Descriptor.pdf). 

>[ Iago Suárez, Ghesn Sfeir, José M. Buenaposada, and Luis Baumela. BEBLID: Boosted efficient binary local image descriptor.
Pattern Recognition Letters, 133:366–372, 2020.](https://www.sciencedirect.com/science/article/pii/S0167865520301252?casa_token=51OQO6-BYywAAAAA:P76uktJ-GIKPSTu6_05B_dJ5mUIiI6gZsZ8ynsuMrj2VYHWt5Ur12l7CumZFJufJGgsUgFM8iJY)

Testing this implementation with the code of the [A-KAZE tutorial](https://docs.opencv.org/4.5.0/db/d70/tutorial_akaze_matching.html), we have found that, detecting 10000 keypoints with ORB and describing with BEBLID obtains 561 inliers (75%) whereas describing with ORB obtains only 493 inliers (63%).

### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [x] I agree to contribute to the project under Apache 2 License.
- [x] To the best of my knowledge, the proposed patch is not based on a code under GPL or other license that is incompatible with OpenCV
- [x] The PR is proposed to proper branch
- [x] There is reference to original bug report and related work
- [x] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [x] The feature is well documented and sample code can be built with the project CMake

```
opencv_extra=master
force_builders=linux,win,mac,docs
```

### CODE REVIEW

File: modules/xfeatures2d/include/opencv2/xfeatures2d.hpp [hunk 1]
```diff
@@ -179,6 +179,55 @@ class CV_EXPORTS_W LATCH : public Feature2D
...
+561 inliers (75%) whereas describing with ORB obtains only 493 inliers (63%).
+*/
+class CV_EXPORTS_W BEBLID : public Feature2D
+{
+public:
+    /**
+     * @brief  Descriptor number of bits, each bit is a boosting weak-learner.
+     * The user can choose between 512 or 256 bits.
+     */
+    CV_WRAP enum
```
Comment (alalek) [inline modules/xfeatures2d/include/opencv2/xfeatures2d.hpp:215]: Add enum name like `DESCRIPTOR_SIZE`

File: modules/xfeatures2d/include/opencv2/xfeatures2d.hpp [hunk 2]
```diff
@@ -179,6 +179,55 @@ class CV_EXPORTS_W LATCH : public Feature2D
     CV_WRAP static Ptr<LATCH> create(int bytes = 32, bool rotationInvariance = true, int half_ssd_size = 3, double sigma = 2.0);
 };
 
+/*
+* BEBLID Descriptor
+*/
+
+/** @brief Class implementing BEBLID (Boosted Efficient Binary Local Image Descriptor)
+ * @cite Suarez2020BEBLID .
```
Comment (alalek) [inline modules/xfeatures2d/include/opencv2/xfeatures2d.hpp:187]: `@cite` is merged to `@brief` without an empty line.

File: modules/xfeatures2d/include/opencv2/xfeatures2d.hpp [hunk 3]
```diff
@@ -179,6 +179,55 @@ class CV_EXPORTS_W LATCH : public Feature2D
...
+    /** @brief Creates the BEBLID descriptor.
+    @param scale_factor Adjust the sampling window around detected keypoints:
+    - <b> 1.00f </b> should be the scale for ORB keypoints
+    - <b> 6.75f </b> should be the scale for SIFT detected keypoints
+    - <b> 6.25f </b> is default and fits for KAZE, SURF detected keypoints
+    - <b> 5.00f </b> should be the scale for AKAZE, MSD, AGAST, FAST, BRISK keypoints
+    @param n_bits Determine the number of bits in the descriptor. Should be either
+     BEBLID::SIZE_512_BITS or BEBLID::SIZE_256_BITS.
+    */
+    CV_WRAP static Ptr<BEBLID> create(float scale_factor, int n_bits = BEBLID::SIZE_512_BITS);
```
Comment (alalek) [inline modules/xfeatures2d/include/opencv2/xfeatures2d.hpp:228]: > int n_bits

Please use enum name instead.

File: modules/xfeatures2d/include/opencv2/xfeatures2d.hpp [hunk 4]
```diff
@@ -179,6 +179,55 @@ class CV_EXPORTS_W LATCH : public Feature2D
     CV_WRAP static Ptr<LATCH> create(int bytes = 32, bool rotationInvariance = true, int half_ssd_size = 3, double sigma = 2.0);
 };
 
+/*
+* BEBLID Descriptor
+*/
+
```
Comment (alalek) [inline modules/xfeatures2d/include/opencv2/xfeatures2d.hpp:185]: Please remove this comment block (it not not needed).

Other similar places will be removed later (its usage is not consistent).

File: modules/xfeatures2d/src/beblid.cpp [hunk 5]
```diff
@@ -0,0 +1,602 @@
...
+// of this distribution and at http://opencv.org/license.html.
+// Author: Iago Suarez <iago.suarez.canosa@alumnos.upm.es>
+
+// Implementation of the article:
+//     Iago Suarez, Ghesn Sfeir, Jose M. Buenaposada, and Luis Baumela.
+//     BEBLID: Boosted Efficient Binary Local Image Descriptor.
+//     Pattern Recognition Letters, 133:366–372, 2020.
+
+#include <bitset>
+#include "precomp.hpp"
```
Comment (alalek) [inline modules/xfeatures2d/src/beblid.cpp:12]: > precomp.hpp

should go first.

File: modules/xfeatures2d/src/beblid.cpp [hunk 6]
```diff
@@ -0,0 +1,602 @@
...
+
+/**
+ * @brief Function that determines if a keypoint is close to the image border.
+ * @param kp The detected keypoint
+ * @param imgSize The size of the image
+ * @param patchSize The size of the normalized patch where the measurement functions were learnt.
+ * @param scaleFactor A scale factor that magnifies the measurement functions w.r.t. the keypoint.
+ * @return true if the keypoint is in the border, false otherwise
+ */
+inline bool isKeypointInTheBorder(const cv::KeyPoint &kp,
```
Comment (alalek) [inline modules/xfeatures2d/src/beblid.cpp:76]: Please use `static` / `static inline` for local functions (here and below)

File: modules/xfeatures2d/src/beblid.cpp [hunk 7]
```diff
@@ -0,0 +1,602 @@
...
+                                  const cv::Size &patchSize = {32, 32},
+                                  float scaleFactor = 1)
+{
+    // This would be the correct measure but since we will compare with half of the size, use this as border size
+    float s = scaleFactor * kp.size / (patchSize.width + patchSize.height);
+    cv::Size2f border(patchSize.width * s * CV_BEBLID_EXTRA_RATIO_MARGIN,
+                      patchSize.height * s * CV_BEBLID_EXTRA_RATIO_MARGIN);
+
+    if (kp.pt.x < border.width || kp.pt.x + border.width >= imgSize.width) return true;
+    if (kp.pt.y < border.height || kp.pt.y + border.height >= imgSize.height) return true;
```
Comment (alalek) [inline modules/xfeatures2d/src/beblid.cpp:87]: > return true;

Consider using dedicated line for this (to get usable code coverage results).

File: modules/xfeatures2d/src/beblid.cpp [hunk 8]
```diff
@@ -0,0 +1,602 @@
...
+ */
+inline void rectifyABWL(const std::vector<ABWLParams> &wlPatchParams,
+                        std::vector<ABWLParams> &wlImageParams,
+                        const cv::KeyPoint &kp,
+                        float scaleFactor = 1,
+                        const cv::Size &patchSize = cv::Size(32, 32))
+{
+    float m00, m01, m02, m10, m11, m12;
+    float s, cosine, sine;
+    uint32_t i;
```
Comment (alalek) [inline modules/xfeatures2d/src/beblid.cpp:108]: Consider placing vars declaration near its usage.

> uint32_t i

`for (size_t i = 0; i < wlPatchParams.size(); i++)` ?

File: modules/xfeatures2d/src/beblid.cpp [hunk 9]
```diff
@@ -0,0 +1,602 @@
...
+    int box_area1, box_area2, width;
+    float sum1, sum2, average1, average2;
+    // Since the integral image has one extra row and col, calculate the patch dimensions
+    frameWidth = integralImage.cols;
+    frameHeight = integralImage.rows;
+
+    // For the first box, we calculate its margin coordinates
+    box1x1 = wlImageParams.x1 - wlImageParams.boxRadius;
+    if (box1x1 < 0) box1x1 = 0;
+    else if (box1x1 >= frameWidth - 1) box1x1 = frameWidth - 2;
```
Comment (alalek) [inline modules/xfeatures2d/src/beblid.cpp:169]: dedicated lines for `if` body (code coverage).

File: modules/xfeatures2d/src/beblid.cpp [hunk 10]
```diff
@@ -0,0 +1,602 @@
...
+    : scale_factor_(scale_factor), patch_size_(32, 32)
+{
+    // Populate wl_params_
+    if (n_bits == SIZE_512_BITS)
+    {
+        // Pre-trained parameters of BEBLID-512 trained in Liberty data set with
+        // a million of patch pairs, 20% positives and 80% negatives
+        wl_params_ =
+            {{24, 18, 15, 17, 6, 13}, {19, 14, 13, 17, 2, 18}, {23, 19, 12, 15, 6, 19}, {24, 14, 16, 16, 6, 11},
+             {16, 15, 12, 16, 1, 12}, {16, 15, 7, 10, 4, 10}, {17, 12, 8, 17, 3, 16}, {24, 12, 11, 17, 7, 19},
```
Comment (alalek) [inline modules/xfeatures2d/src/beblid.cpp:299]: Please move these blobs into dedicated files (to keep source code clean):
- `beblid.p512.hpp`
```
// Pre-trained parameters of BEBLID-512 trained in Liberty data set with
// a million of patch pairs, 20% positives and 80% negatives
static const ABWLParams wl_params_512[] = {
    {24, 18, 15, 17, 6, 13}, {19, 14, 13, 17, 2, 18}, {23, 19, 12, 15, 6, 19}, {24, 14, 16, 16, 6, 11},
    ...
};
```
- `beblid.p256.hpp` with similar content


Use them:
```
{
    #include "beblid.p512.hpp"
    #include "beblid.p256.hpp"
    if (n_bits == SIZE_512_BITS)
        wl_params_.assign(wl_params_512, wl_params_512 + sizeof(wl_params_512) / sizeof(wl_params_512[0]));
    else if(n_bits == SIZE_256_BITS)
        wl_params_.assign(wl_params_256, wl_params_256 + sizeof(wl_params_256) / sizeof(wl_params_256[0]));
    else
        CV_Error(...);
}
```

File: modules/xfeatures2d/src/beblid.cpp [hunk 11]
```diff
@@ -0,0 +1,602 @@
...
+    const int *integralPtr = integralImg.ptr<int>();
+    cv::Size frameSize(integralImg.cols - 1, integralImg.rows - 1);
+
+    // Parallel Loop to process descriptors
+#ifndef CV_BEBLID_PARALLEL
+    const cv::Range range(0, keypoints.size());
+#else
+    cv::parallel_for_(cv::Range(0, int(keypoints.size())), [&](const Range &range)
+    {
+#endif
```
Comment (alalek) [inline modules/xfeatures2d/src/beblid.cpp:518]: BTW, You may swap `{` and `#endif` lines here and use this below

```
    }
#ifdef CV_BEBLID_PARALLEL
    );
#endif
```

**Review by alalek — COMMENTED** [review]
Great contribution!

Please take a look on comments below.

File: modules/xfeatures2d/include/opencv2/xfeatures2d.hpp (hunk 1 above)
Comment (iago-suarez) [reply modules/xfeatures2d/include/opencv2/xfeatures2d.hpp]: I have added a new enum **`BeblidSize`**. It gave me some problem with the python wrappers, so I imitated the Daisy strategy adding the enum to the files [`gen_dict.json`](https://github.com/opencv/opencv_contrib/pull/2774/commits/0f10bd3ff16c6eb285dda2dbabd36675de544d10#diff-acd9128554f80b55f5448d9e6a7b4769156f30911e06b955a012def9ea1eb9d7) and [`pyopencv_xfeatures2d.hpp`](https://github.com/opencv/opencv_contrib/pull/2774/commits/0f10bd3ff16c6eb285dda2dbabd36675de544d10#diff-ca37445c0bbe1a70f0642d66ddcbfcb5dc6faf2a7cecfe3bf1a5893e3a2d1b15). Please, let me known if this solution is valid or there is a better way.

File: modules/xfeatures2d/include/opencv2/xfeatures2d.hpp (hunk 2 above)
Comment (iago-suarez) [reply modules/xfeatures2d/include/opencv2/xfeatures2d.hpp]: I fixed it with the phrase ", described in [XXX]" as other descriptors do.

File: modules/xfeatures2d/include/opencv2/xfeatures2d.hpp (hunk 3 above)
Comment (iago-suarez) [reply modules/xfeatures2d/include/opencv2/xfeatures2d.hpp]: done

File: modules/xfeatures2d/include/opencv2/xfeatures2d.hpp (hunk 4 above)
Comment (iago-suarez) [reply modules/xfeatures2d/include/opencv2/xfeatures2d.hpp]: done

File: modules/xfeatures2d/src/beblid.cpp (hunk 5 above)
Comment (iago-suarez) [reply modules/xfeatures2d/src/beblid.cpp:11]: You are right, in fact `<bitset>` is no used so I deleted it.

File: modules/xfeatures2d/src/beblid.cpp (hunk 6 above)
Comment (iago-suarez) [reply modules/xfeatures2d/src/beblid.cpp]: done

File: modules/xfeatures2d/src/beblid.cpp (hunk 7 above)
Comment (iago-suarez) [reply modules/xfeatures2d/src/beblid.cpp]: done

File: modules/xfeatures2d/src/beblid.cpp (hunk 8 above)
Comment (iago-suarez) [reply modules/xfeatures2d/src/beblid.cpp]: done

File: modules/xfeatures2d/src/beblid.cpp (hunk 9 above)
Comment (iago-suarez) [reply modules/xfeatures2d/src/beblid.cpp]: done, I have fixed this problem also other lines where the body was in the same line. Thank you

File: modules/xfeatures2d/src/beblid.cpp (hunk 10 above)
Comment (iago-suarez) [reply modules/xfeatures2d/src/beblid.cpp]: Done. I located the files in [`opencv_contrib/modules/xfeatures2d/src/beblid.p512.hpp`](https://github.com/opencv/opencv_contrib/pull/2774/commits/0f10bd3ff16c6eb285dda2dbabd36675de544d10#diff-78494b0c74baff117f619067e1a8637dac3686f54ea8101ca146f8e9e06fa90c) and [`opencv_contrib/modules/xfeatures2d/src/beblid.p256.hpp`](https://github.com/opencv/opencv_contrib/pull/2774/commits/0f10bd3ff16c6eb285dda2dbabd36675de544d10#diff-06223661d4e54622d84f21aee4c0ec65b339e072d1b6044c0b11ea1c2e75bf1e) please let me known if this is the correct place.

File: modules/xfeatures2d/src/beblid.cpp (hunk 11 above)
Comment (iago-suarez) [reply modules/xfeatures2d/src/beblid.cpp]: Good point! It's done.

File: modules/xfeatures2d/src/beblid.cpp (hunk 10 above)
Comment (alalek) [reply modules/xfeatures2d/src/beblid.cpp]: Location is correct.

Please remove `#ifndef __OPENCV_XFEATURES2D_BEBLID_P256_HPP__` guards from them. These headers are not expected to be used as "standalone".

Comment (iago-suarez) [reply modules/xfeatures2d/src/beblid.cpp]: Done! Thank you again.

File: modules/xfeatures2d/include/opencv2/xfeatures2d.hpp [hunk 12]
```diff
@@ -179,6 +179,52 @@ class CV_EXPORTS_W LATCH : public Feature2D
...
+    @param scale_factor Adjust the sampling window around detected keypoints:
+    - <b> 1.00f </b> should be the scale for ORB keypoints
+    - <b> 6.75f </b> should be the scale for SIFT detected keypoints
+    - <b> 6.25f </b> is default and fits for KAZE, SURF detected keypoints
+    - <b> 5.00f </b> should be the scale for AKAZE, MSD, AGAST, FAST, BRISK keypoints
+    @param n_bits Determine the number of bits in the descriptor. Should be either
+     BEBLID::SIZE_512_BITS or BEBLID::SIZE_256_BITS.
+    */
+    CV_WRAP static Ptr<BEBLID> create(float scale_factor,
+                                      BEBLID::BeblidSize n_bits = BEBLID::SIZE_512_BITS);
```
Comment (alalek) [inline modules/xfeatures2d/include/opencv2/xfeatures2d.hpp:225]: OK, lets use `int` here for now.

I filled Java binding generator bug here: https://github.com/opencv/opencv/issues/19056

File: modules/xfeatures2d/src/beblid.cpp [hunk 13]
```diff
@@ -0,0 +1,423 @@
...
+    // Calculate the indices on the integral image where the box falls
+    width = integralImage.cols;
+    idx1 = box1y1 * width + box1x1;
+    idx2 = box1y1 * width + box1x2;
+    idx3 = box1y2 * width + box1x1;
+    idx4 = box1y2 * width + box1x2;
+    CV_DbgAssert(idx1 >= 0 && idx1 < integralImage.size().area());
+    CV_DbgAssert(idx2 >= 0 && idx2 < integralImage.size().area());
+    CV_DbgAssert(idx3 >= 0 && idx3 < integralImage.size().area());
+    CV_DbgAssert(idx4 >= 0 && idx4 < integralImage.size().area());
```
[the remark this answers is not in the document]
Comment (iago-suarez) [reply modules/xfeatures2d/src/beblid.cpp]: done, the code is now much easier to read and the number of operations remains the same. Thanks a bunch!

File: modules/xfeatures2d/src/beblid.cpp [hunk 14]
```diff
@@ -0,0 +1,423 @@
...
+// descriptor computation using keypoints
+void BEBLID_Impl::compute(InputArray _image, vector<KeyPoint> &keypoints, OutputArray _descriptors)
+{
+    Mat image = _image.getMat();
+
+    if (image.empty())
+        return;
+
+    if (keypoints.empty())
+        return;
```
[the remark this answers is not in the document]
Comment (iago-suarez) [reply modules/xfeatures2d/src/beblid.cpp:253]: done

File: modules/xfeatures2d/include/opencv2/xfeatures2d.hpp (hunk 12 above)
Comment (iago-suarez) [reply modules/xfeatures2d/include/opencv2/xfeatures2d.hpp]: Perfect, I changed to use `int` again

**Review by alalek — APPROVED** [review]
Well done! Thank you for contribution :+1:

**Comment by iago-suarez** [comment]
Thanks for your review and comments @alalek , it's a pleasure to contribute! :smiley:

**Pull request merged**